libs/contrib/Data/Vect/Sort.idr

```diff
@@ -0,0 +1,39 @@
+module Data.Vect.Sort
+
+import Data.Vect
+import Data.Nat
+
+-- Merge sort implementation for Vect n a
+export total
+sortBy : {n : Nat} -> (cmp : a -> a -> Ordering) -> (xs : Vect n a) -> Vect n a
+sortBy cmp [] = []
+sortBy cmp [x] = [x]
+sortBy cmp xs =
+    let (p ** (q ** equalityProof)) = div2 n in
+        let (left, right) = splitAt p (pqRewrite equalityProof xs) in
+            rewrite sym equalityProof in
+                mergeBy cmp
+                    (sortBy cmp (assert_smaller xs left))
+                    (sortBy cmp (assert_smaller xs right))
+        where
+            pqRewrite : (p + q = n) -> Vect n a -> Vect (p + q) a
+            pqRewrite equalityProof xs = rewrite equalityProof in xs
+
+            -- Divides n by 2 (rounding down), returning `p = n / 2` and
+            -- `q = n - (n / 2)`, along with a proof that `p + q = n`.
+            div2 : (n : Nat) -> (p ** (q ** (p + q = n)))
```

Comment:
This is worth lifting to the toplevel.

Reply:
Sure, I can certainly do that. Do you think it should go in this module, or one related to `Nat`?

Reply:
I would put it in the `Nat` one.

Reply:
The comment there sounds like it is making guarantees about the numbers returned. Could those guarantees be incorporated into the signature? Otherwise this claim admits of a trivial proof:

```idris
div2 : (n : Nat) -> (p ** (q ** (p + q = n)))
div2 n = (Z ** (n ** Refl))
```

Reply:
Agreed @nickdrozd I want to include something along the lines of `(p = q) or (p = S q)` (depending on if `n` is even or odd of course), in addition to the `p + q = n` constraint. I wasn't sure how to cleanly encode it into the signature though. I was wondering if I should include an additional lemma, aside from the main definition which proves this fact.

Reply:
@birchmd Sounds like you want the [half](https://github.com/idris-lang/Idris2/blob/master/libs/base/Data/Nat/Views.idr#L19) function.

Reply:
@gallais Yes, that is exactly what I wanted! Thanks for the tip, I didn't know that was in the base library. Please see the updated implementation using `half`. https://github.com/idris-lang/Idris2/pull/417/commits/d5eb34489286e99657a1fbafe4e79fa2b5e32f97